File: internal/model/menu.go
package model

type MenuType string

const (
	MenuTypeDirectory MenuType = "directory"
	MenuTypeMenu      MenuType = "menu"
	MenuTypeButton    MenuType = "button"
)

type Menu struct {
	BaseModel
	ParentID   *uint    `json:"parentId" gorm:"index"`
	Name       string   `json:"name" gorm:"size:64;not null"`
	Type       MenuType `json:"type" gorm:"size:16;not null"`
	Path       string   `json:"path" gorm:"size:255"`
	Icon       string   `json:"icon" gorm:"size:64"`
	Permission string   `json:"permission" gorm:"size:128;uniqueIndex:idx_permission_nonempty"`
	Sort       int      `json:"sort" gorm:"default:0"`
	IsHidden   bool     `json:"isHidden" gorm:"not null;default:false"`
	Children   []Menu   `json:"children,omitempty" gorm:"foreignKey:ParentID"`
}

File: web/src/stores/menu.ts
import { create } from "zustand"
import { api } from "@/lib/api"

export interface MenuItem {
  id: number
  parentId: number | null
  name: string
  type: "directory" | "menu" | "button"
  path: string
  icon: string
  permission: string
  sort: number
  isHidden: boolean
  children: MenuItem[]
}

interface MenuState {
  menuTree: MenuItem[]
  permissions: string[]
  initialized: boolean

  init: () => Promise<void>
  clear: () => void
  hasPermission: (code: string) => boolean
}

export const useMenuStore = create<MenuState>((set, get) => ({
  menuTree: [],
  permissions: [],
  initialized: false,

  init: async () => {
    try {
      const data = await api.get<{ menus: MenuItem[]; permissions: string[] }>(
        "/api/v1/menus/user-tree",
      )
      set({
        menuTree: data.menus || [],
        permissions: data.permissions || [],
        initialized: true,
      })
    } catch {
      set({ menuTree: [], permissions: [], initialized: true })
    }
  },

  clear: () => {
    set({ menuTree: [], permissions: [], initialized: false })
  },

  hasPermission: (code: string) => {
    return get().permissions.includes(code)
  },
}))
